consensus/state_sync/module.go

```diff
@@ -24,41 +43,140 @@ var (
-	// when all blocks are received and committed, stops the state sync process by calling its `Stop()` function.
+	go m.metadataSyncLoop()
 	return nil
 }
 
-// TODO(#352): check if node is a valdiator, if not send Consensus_IsSyncedNonValidator event
-// Stop stops the state sync process, and sends `Consensus_IsSyncedValidator` FSM event
-func (m *stateSync) Stop() error {
+// Start a synchronous state sync process to catch up to the network
+// 1. Processes and aggregates all metadata collected in metadataReceived channel
+// 2. Requests missing blocks until the maximum seen block is retrieved
+// 3. Perform (2) one-by-one, applying and validating each block while doing so
+// 4. Once all blocks are received and committed, stop the synchronous state sync process
+func (m *stateSync) StartSynchronousStateSync() error {
+	consensusMod := m.bus.GetConsensusModule()
+	currentHeight := consensusMod.CurrentHeight()
+	nodeAddress := consensusMod.GetNodeAddress()
+	nodeAddressBz, err := hex.DecodeString(nodeAddress)
+	if err != nil {
+		return err
+	}
+
+	readCtx, err := m.GetBus().GetPersistenceModule().NewReadContext(int64(currentHeight))
+	if err != nil {
+		return err
+	}
+	defer readCtx.Release()
+
+	// Get a view into the state of the network
+	_, maxHeight := m.getAggregatedStateSyncMetadata()
+
+	// Synchronously request block requests from the current height to the aggregated metadata height
+	// Note that we are using `<=` because:
+	// - maxHeight is the max * committed * height of the network
+	// - currentHeight is the latest * committing * height of the node
+	for currentHeight <= maxHeight {
+		m.logger.Info().Msgf("Synchronous state sync is requesting block: %d, ending height: %d", currentHeight, maxHeight)
+
+		// form the get block request message
+		stateSyncGetBlockMsg := &typesCons.StateSyncMessage{
+			Message: &typesCons.StateSyncMessage_GetBlockReq{
+				GetBlockReq: &typesCons.GetBlockRequest{
+					PeerAddress: nodeAddress,
+					Height:      currentHeight,
+				},
+			},
+		}
+		anyProtoStateSyncMsg, err := anypb.New(stateSyncGetBlockMsg)
+		if err != nil {
+			return err
+		}
+
+		// Broadcast the block request
+		if err := m.GetBus().GetP2PModule().Broadcast(anyProtoStateSyncMsg); err != nil {
+			return err
+		}
+
+		// Wait for the consensus module to commit the requested block and re-try on timeout
+		select {
+		case blockHeight := <-m.committedBlocksChannel:
+			m.logger.Info().Msgf("State sync received event that block %d is committed!", blockHeight)
+		case <-time.After(blockWaitingPeriod):
+			m.logger.Warn().Msgf("Timed out waiting for block %d to be committed...", currentHeight)
+		}
+
+		// Update the height and continue catching up to the latest known state
+		currentHeight = consensusMod.CurrentHeight()
+	}
+
+	// Checked if the synched node is a validator or not
+	isValidator, err := readCtx.GetValidatorExists(nodeAddressBz, int64(currentHeight))
```

Comment:
Why do `PersistenceReadContext` most of the methods take a `height` when a `height` is required to get the context in the first place? Are they not necessarily the same in some scenarios?

Reply:
https://github.com/pokt-network/pocket/issues/406

Reply:
Thanks for the additional context!